### PR TITLE
Edit Trip Modal Infrastructure

### PR DESCRIPTION
**Pull request by zghera**
### What is a quick description of the change?

Set up "infrastructure" such that the `SaveTripModal` does the work of the current `AddTripModal` and what would be a `EditTripModal`.

### Is this fixing an issue?

none

### Are there more details that are relevant?

The Edit Trip modal was split up into two PRs, this one and #73 to avoid one very large PR with potential issues when it came to renaming and making many changes on that renamed file.

`SaveTripModal` replaces `AddTripModal` and does the same work expect now has three additional props: `title`, `tripId`, and `placeholderObj` that determine whether the modal will be responsible for adding a new trip or editing an existing trip. 

### Check lists (check `x` in `[ ]` of list items)

- [  ] Test written/updating
- [  ] Tests passing
- [x] Coding style (indentation, etc)

No unit tests as the only changes are infrastructural/organizational.

As of now, integration (end-to-end) tests were deemed non-critical. Thus, functions related to react and firestore will be tested at a later time. Unit tests for "logical" functions will be included in this PR and for the remaining PRs needed to complete the MVP.

### CODE REVIEW

File: frontend/src/components/ViewTrips/create-new-trip.js [hunk 1]
```diff
@@ -96,26 +96,36 @@ function formatTripData(rawTripObj) {
  * Promise containing a reference to the newly created document.
  *
  * @param {firebase.firestore.Firestore} db The Firestore database instance.
+ * @param {?string} tripId Document id associated with trip to overwrite.
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/create-new-trip.js:99]: What does ?string do?

File: frontend/src/components/ViewTrips/create-new-trip.js [hunk 2]
```diff
@@ -96,26 +96,36 @@ function formatTripData(rawTripObj) {
...
  * @param {Object} tripObj A JS Object containing the Trip document fields.
  */
-export function addTripToFirestore(db, tripObj) {
-  return db.collection(COLLECTION_TRIPS)
-    .add(tripObj)
+export function addTripToFirestore(db, tripId, tripObj) {
+  if (tripId === null) {
+    return db.collection(COLLECTION_TRIPS).add(tripObj);
+  }
+  return db.collection(COLLECTION_TRIPS).doc(tripId).set(tripObj);
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/create-new-trip.js:107]: In what instance would the tripId already exist? How do we get to this line?

File: frontend/src/components/ViewTrips/create-new-trip.js [hunk 3]
```diff
@@ -96,26 +96,36 @@ function formatTripData(rawTripObj) {
...
 }
 
 /**
  * Formats/cleans form data and creates new Trip document in firestore.
  *
  * @param {firebase.firestore.Firestore} db The Firestore database instance.
+ * @param {?string} tripId Document id associated with trip to overwrite.
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/create-new-trip.js:114]: Why should we be able to overwrite a trip? Can a user do this intentionally?

File: frontend/src/components/ViewTrips/index.js [hunk 4]
```diff
@@ -19,13 +19,23 @@ class ViewTrips extends React.Component {
     super();
     this.state = { showModal: false,
                    refreshTripsContainer: false,
-                   refreshAddTripModal: false
+                   refreshSaveTripModal: false,
+                   modalTitle: null,
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/index.js:23]: Is modalTitle used somewhere? I can't find it..

File: frontend/src/components/ViewTrips/create-new-trip.js (hunk 2 above)
Comment (ananyeet) [reply frontend/src/components/ViewTrips/create-new-trip.js:107]: If you're trying to update a trip, try `.doc().update(tripObj)`. It changes the fields provided without erasing any of the other fields.

File: frontend/src/components/ViewTrips/save-trip-modal.js [hunk 5]
```diff
@@ -98,38 +99,59 @@ function createFormGroup(controlId, formLabel, inputType, placeholder, ref) {
...
+ * the two fake components are dervied from the props `title`, `tripid`, and
+ * `placeholderObj` (see below).
  *
  * @param {Object} props These are the props for this component:
  * - db: Firestore database instance.
  * - show: Boolean that determines if the add trips modal should be displayed.
  * - handleClose: The function that handles closing the add trips modal.
  * - refreshTripsContainer: Function that handles refreshing the TripsContainer
  *        component upon trip creation (Remove when fix Issue #62).
+ * - title: The title of the modal.
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/save-trip-modal.js:126]: What would this title be? Is it something that would be required to be passed in? i.e. if the title is "edit trip" vs "create new trip", can you check if tripId==null and set the title that way? 
This is fine too I guess, it's just easier to follow if there are fewer black boxes.

File: frontend/src/components/ViewTrips/save-trip-modal.js [hunk 6]
```diff
@@ -98,38 +99,59 @@ function createFormGroup(controlId, formLabel, inputType, placeholder, ref) {
...
+ * - tripId: For adding a new trip, this will be null. For editting an existing
+ *        trip, this will the document id associated with the trip.
+ * - placeholderObj: Object containing the placeholder/default values for the
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/save-trip-modal.js:129]: same as above with placeholderobj. Can you pull stuff out of the database when tripId != null?

File: frontend/src/components/ViewTrips/save-trip-modal.js [hunk 7]
```diff
@@ -98,38 +99,59 @@ function createFormGroup(controlId, formLabel, inputType, placeholder, ref) {
...
+ *
+ * This component "acts" as a parent of the (non-existant) AddTripModal and
+ * EditTripModal components. The only differences in the implementation between
+ * the two fake components are dervied from the props `title`, `tripid`, and
+ * `placeholderObj` (see below).
  *
  * @param {Object} props These are the props for this component:
  * - db: Firestore database instance.
  * - show: Boolean that determines if the add trips modal should be displayed.
  * - handleClose: The function that handles closing the add trips modal.
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/save-trip-modal.js:123]: curious- What are the potential functions handleClose can be?

File: frontend/src/components/ViewTrips/create-new-trip.js (hunk 1 above)
Comment (zghera) [reply frontend/src/components/ViewTrips/create-new-trip.js:99]: It means that it is a nullable type: https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System.

File: frontend/src/components/ViewTrips/create-new-trip.js (hunk 2 above)
Comment (zghera) [reply frontend/src/components/ViewTrips/create-new-trip.js]: So they way I envisioned the 'Edit Trip' modal is similar to the Add Trip modal except each text box is populated with the current values of the trip rather than being blank with the placeholder input values. So if there were 10 collaborators, rather than having to re-enter them all, they can choose to delete one and add two and keep the other nine the same. With this model, it doesn't really matter whether I overwrite with `set` or update with `update` because I will get the same result (and I doubt there is a significant performance difference. And because I see there are a few complications with update (such as updating arrays), I think it makes sense for me to stay with using `set`.

If you disagree just let me know and we can chat about it.

File: frontend/src/components/ViewTrips/save-trip-modal.js (hunk 7 above)
Comment (zghera) [reply frontend/src/components/ViewTrips/save-trip-modal.js:123]: Only `hideSaveTripModal`

File: frontend/src/components/ViewTrips/save-trip-modal.js (hunk 5 above)
Comment (zghera) [reply frontend/src/components/ViewTrips/save-trip-modal.js:126]: Yes this one I could pretty easily do inside of save-trip-modal so I can remove this prop.

File: frontend/src/components/ViewTrips/save-trip-modal.js (hunk 6 above)
Comment (zghera) [reply frontend/src/components/ViewTrips/save-trip-modal.js:129]: This one is not so easy. I did not do it in this PR (in PR #73 but not ready for review just yet) but I have to pass a event handler down from `index.js` to trip.js` so when the 'Edit Trip' button is pressed, the edit trip modal is opened. So if I were to remove this I would have to somehow have a handler in index that would somehow have to reference I handler in save-trip-modal.js. It will be clearer when you see it but ...

tldr: This way will be a lot cleaner based on my current knowledge.

File: frontend/src/components/ViewTrips/save-trip-modal.js [hunk 8]
```diff
@@ -110,29 +111,48 @@ function createFormGroup(controlId, formLabel, inputType, placeholder, ref) {
 }
 
 /**
- * Component corresponding to add trips modal.
+ * Component corresponding to the save trips modal.
+ *
+ * This component "acts" as a parent of the (non-existant) AddTripModal and
```
[the remark this answers is not in the document]
Comment (zghera) [reply frontend/src/components/ViewTrips/save-trip-modal.js:114]: Fixed.

File: frontend/src/components/ViewTrips/index.js [hunk 9]
```diff
@@ -36,36 +45,84 @@ class ViewTrips extends React.Component {
...
+  showAddTripModal = () => {
+    this.setState({
+      tripId: null,
+      placeholderObj: {
+        name:          'Enter Trip Name',
+        description:   'Enter Trip Description',
+        destination:   'Enter Trip Destination',
+        startDate:     '',
+        endDate:       '',
+        collaborators: ['person@email.xyz']
```
[the remark this answers is not in the document]
Comment (zghera) [reply frontend/src/components/ViewTrips/index.js:84]: No it is supposed to hold the placeholder for now (this is changed in #73). The 'infrastructure' is slightly changed in #73 but this idea was that this value would be seen in the email field when creating a new trip. And when editing an existing trip, the existing collaborator email will show up rather than the placeholder.
